ann_benchmarks/algorithms/faiss_hnsw.py

```diff
@@ -0,0 +1,33 @@
+from __future__ import absolute_import
+import os
+import faiss
+import numpy as np
+from ann_benchmarks.constants import INDEX_DIR
+from ann_benchmarks.algorithms.base import BaseANN
+
+
+class FaissHNSW(BaseANN):
+    def __init__(self, metric, method_param):
+        self.metric = metric
+        self.method_param = method_param
+        self.name = 'faiss (%s)' % (self.method_param)
+
+    def fit(self, X):
+
+        self.index = faiss.IndexHNSWFlat(len(X[0]),self.method_param["M"])
+        self.index.hnsw.efConstruction = self.method_param["efConstruction"]
+        self.index.verbose = True
+        if(self.metric == 'angular'):
+            X = X / np.linalg.norm(X, axis=1)[:, np.newaxis]
+        self.index.add(X)
+        faiss.omp_set_num_threads(1)
```

Comment:
i thought the cpu affinity code already takes care of this?

Reply:
Affinity makes sure that no one takes advantage of multiply threads during the query phase. 
AFAIK openmp and other libs are not aware of the externally set affinity and thus would try to start many threads leading to performance loss.
Faiss uses all cores by default, so probably it is a good idea to set the number of threads to 1.

Reply:
well it's not a fair comparison if some libraries use multiple threads and some libraries don't

Should we remove the CPU affinity and enable an optional thread pool for queries?

Reply:
Seems fair to me. 
IMHO having a single-thread regime is obligatory for such libraries.
And if some lib uses multiply threads by default it should be disabled manually in the wrapper. Does not seem like a big deal.

Reply:
Well the problem is when OpenMP/Eigen uses threads under the hood, then it might be hard to fully control how many threads are being used.

I can make it configurable to use multithreading (or multiprocessing) for libraries (Annoy would benefit from this for instance). Would probably enable this by default and opt-out for libraries that implement their own threading.

Reply:
I'm against supporting batch mode since it would make it hard to compare algorithm benchmarks across libraries.

In my real world experience (mostly using Annoy at Spotify), very few use cases are compatible with batch queries

Reply:
> I do not see how threadpool alone will solve this problem. Some algorithms may still create a lot useless threads (in this case num_threads**2).

That's why you make it configurable. So some algorithms will use their own internal threading, so there's no need to use the thread pool (or multiprocessing). Some algorithms do not use threads under the hood so they will benefit from a thread/multiprocessing pool.

I don't have super strong feelings about what the default behavior should be. Could be opt-in or opt-out (either way Annoy would use it and NMSLIB would not, etc)

Reply:
I agree that batch mode is somewhat limited, but there are cases where it is useful.
For instance, it is obviously used in production in Facebook; I know lots of companies use batch mode for kNN on GPUs for processing non-realtime data; K-Means and other clusterization algorithms also may benefit from KNN in batch mode.

I am also a bit worrying about the performance measuring issues, like   https://github.com/erikbern/ann-benchmarks/issues/47 . However this might not be the case for multithreading and now can be monitored by looking at the results of the "dummy" algorithms.

IMO single thread performance is a good common denominator :)

Reply:
i'd like to stick to single thread, but somewhat concerned not all libraries will respect running on a single thread.

but i can run the benchmarks and keep an eye on the cpu usage to make sure they don't use multiple cpus

actually looks like i can do it using docker: https://docs.docker.com/config/containers/resource_constraints/ – i'll try that!